Fix/invalidate whats new cache on version change

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes: [AINFRA-1635: Feature announcements are never refreshed when app version changes](https://linear.app/a8c/issue/AINFRA-1635/feature-announcements-are-never-refreshed-when-app-version-changes)

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR fixes a bug in which, after initial fetch, the app never again requestes "what's new" announcements. 

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->

<details><summary>Clean Shared Preferences <code>LAST_VERSION_WITH_ANNOUNCEMENT</code></summary>

```shell
adb shell "run-as com.woocommerce.android.dev sed -i '/<string name=\"LAST_VERSION_WITH_ANNOUNCEMENT\"/d' /data/data/com.woocommerce.android.dev/shared_prefs/com.woocommerce.android.dev_preferences.xml" && adb shell am force-stop com.woocommerce.android.dev
```

</details> 

<details><summary>API Faker payload</summary>

```json
[
  {
    "request": {
      "httpMethod": "GET",
      "id": 0,
      "path": "/wpcom/v2/mobile/feature-announcements",
      "queryParameters": [],
      "type": {
        "type": "wp-com"
      }
    },
    "response": {
      "body": "{\n    \"announcements\": [\n      {\n        \"appVersionName\": \"19.8\",\n        \"announcementVersion\": 1,\n        \"minimumAppVersion\": \"1.0\",\n        \"maximumAppVersion\": \"99.0\",\n        \"appVersionTargets\": [\"19.0\"],\n        \"detailsUrl\": \"\",\n        \"isLocalized\": true,\n        \"responseLocale\": \"en\",\n        \"features\": [\n          {\n            \"title\": \"New Feature Title\",\n            \"subtitle\": \"Description of the new feature\",\n            \"iconBase64\": \"\",\n            \"iconUrl\": \"https://s0.wordpress.com/i/store/mobile/plans-premium.png\"\n          },\n          {\n            \"title\": \"Another Feature\",\n            \"subtitle\": \"Another cool feature description\",\n            \"iconBase64\": \"\",\n            \"iconUrl\": \"https://s0.wordpress.com/i/store/mobile/plans-premium.png\"\n          }\n        ]\n      }\n    ]\n  }",
      "endpointId": 0,
      "statusCode": 200
    }
  }
]
```

</details> 

#### Reproduction

1. In `com.woocommerce.android.util.BuildConfigWrapper` set `versionName` to any constant value, e.g. `19.0`
1. Install app from `trunk`
1. Import the above payload to API Faker
2. Close the app
3. Run `Clean Shared Preferences` from the snippet above (just to have a clean state before tests)
4. Open the app
  - ✅ Assert that you can see "What's new" announcement
5. Go to API Faker settings
6. Update the payload e.g. rename "New Feature Title" to "Test"
7. Save the updated payload
8. In Android Studio, increase `BuildConfigWrapper#versionName` (to simulate new app installation)
9. Install the app
  - ❌ See that "What's New in WooCommerce" shows OLD values

#### Test fix

Repeat above steps but on this branch and see that at the end you see the updated values.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Introduce unit tests for the expected What's New behavior

- if app version changes, we should fetch announcements again
- if app version doesn't change, we should not fetch/query announcements
- if app version changes, but there are no new announcements, we should update preferences to not fetch data on the next run

wzieba committed:
Fix not fetching announcement on app upgrade

Now, if version of the latest version with shown announcement doesn't match current version, the app will fetch new announcement and show it if needed.

Otherwise, it'll just skip the method.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt:
```diff
@@ -227,15 +227,15 @@ class MainActivityViewModel @Inject constructor(
+            }
+
+            val announcement = featureAnnouncementRepository.getLatestFeatureAnnouncement(fromCache = false)
+            prefs.setLastVersionWithAnnouncement(buildConfigWrapper.versionName)
```
🧇 This seems a bit off to me, are you sure about this change?
1. We invoke setLastVersionWithAnnouncement even if the request to fetch the announcement fails.
2. We invoke setLastVersionWithAnnouncement also in the FeatureAnnouncementViewModel when it's shown, which feels more appropriate.

Wdyt?

wzieba replied:
Thanks! Sure, we can remove `prefs.setLastVersionWithAnnouncement(buildConfigWrapper.versionName)` from here to address the case when `getLatestFeatureAnnouncement` would fail, good idea 👍

wzieba replied:
> We invoke setLastVersionWithAnnouncement also in the FeatureAnnouncementViewModel when it's shown, which feels more appropriate.

The problem with this is that if the backend doesn't offer announcements for the specific version of the app, then we'll never call `setLastVersionWithAnnouncement` (because `FeatureAnnouncementViewModel` won't ever start).

This means that each time the app would make an unnecessary HTTP request on start.

wzieba replied:
Anyway, I've removed this line. My intention here was to optimize so the app doesn't call backend if there are no new announcements for this version of the app, but covering the case of a failed http request that you mentioned would require a refactor to `Result` from repository (because now we get `null` if there are no announcements OR when there's an error).

malinajirka reviewed (commented):
Thanks @wzieba for working on this!

I have very little context about how the announcement works, and tbh I'm a bit confused. 

I thought each announcement is displayed just once for each user. I also thought that each announcement has its version and it'd be displayed again only if we updated the version of the announcement.

But if I understand the code correctly, we display the announcement anytime the app is updated, given the min/max announcement versions match? It also seems that the announcement version is completely ignored?

What am I missing?

wzieba commented:
thanks for taking a look @malinajirka !

> I thought each announcement is displayed just once for each user. 

Yes, this is my understanding as well. But currently on `trunk`, announcement will be displayed **once per app lifetime**. If user sees an announcement for version `19.0`, it won't see any announcement for version `20.0`, `21.0` etc. Maybe the reproduction instruction could bring more light

> I also thought that each announcement has its version

It has, but it's not used in the app anywhere. I remove it in a different PR: https://github.com/woocommerce/woocommerce-android/pull/15075/changes/1c41d3d18760ca62047b09cb5cc528b0988a1148#diff-fea7c31ec2218d75212182e1ceb6aa44d5d54bb54e0578b4cb2fbede134b9e37L10

> it'd be displayed again only if we updated the version of the announcement.

The announcement is when the version of the app is between `minimumAppVersion` and `maximumAppVersion` + a few more cases (see [`FeatureAnnouncement#canBeDisplayedOnAppUpgrade`](https://github.com/woocommerce/woocommerce-android/blob/e29e56f60989dd35f225cc0ae99b33449950e4fc/WooCommerce/src/main/kotlin/com/woocommerce/android/model/FeatureAnnouncement.kt#L27))

> But if I understand the code correctly, we display the announcement anytime the app is updated, given the min/max announcement versions match? It also seems that the announcement version is completely ignored?

I didn't change the logic on _if_ to display the announcement. I just make sure, that new announcement is fetched whenever app gets updated.

wzieba committed:
Remove setting `setLastVersionWithAnnouncement` after the `getLatestFeatureAnnouncement(fromCache = false)`

This was an optimization that would save us from an additional HTTP request on the app start.

It's not necessary though, and brings a risk of not ever showing announcement if `getLatestFeatureAnnouncement` fails. Addressing "failed request" case would need a refactor to distinguish between `null` (no announcements) and `failure` (http request failed).

So we keep things as they're now on `trunk` - each time on the app start the HTTP request is triggered.

wzieba committed:
Move `showFeatureAnnouncementIfNeeded` to Activity's `onCreate`

Now that `showFeatureAnnouncementIfNeeded` fetched data from remote, we no longer need to request data update on ViewModel's init.

It was broken anyway, because before `featureAnnouncementRepository.getFeatureAnnouncements(fromCache = false)` had a chance to finish, the `showFeatureAnnouncementIfNeeded` with its `getFeatureAnnouncements(fromCache = true)` (before the fix in previous commit), showed cached data.
`

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivityViewModel.kt:
```diff
@@ -227,15 +227,15 @@ class MainActivityViewModel @Inject constructor(
-                if (prefs.getLastVersionWithAnnouncement() != buildConfigWrapper.versionName &&
-                    cachedAnnouncement.canBeDisplayedOnAppUpgrade(buildConfigWrapper.versionName)
-                ) {
+            if (prefs.getLastVersionWithAnnouncement() == buildConfigWrapper.versionName) {
+                return@launch
+            }
```
In case it'll help, it's worth noting that this condition _is not new_. It's just moved from the lines below. This is a slight optimization to not fetch announcements if they won't be shown anyway.

wzieba committed:
Adjust names of tests to match implementation

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/main/MainActivityViewModelTest.kt:
```diff
@@ -337,9 +339,11 @@ class MainActivityViewModelTest : BaseUnitTest() {
 
     @Test
-    fun `given existing announcement cache, when app is upgraded and announcement is valid, track event is tracked`() =
+    fun `given announcement is available on remote, when app is upgraded and announcement is valid, track event is tracked`() =
```
🔍 The last part should be `when track ....` - otherwise detekt will complain.

malinajirka reviewed (approved):
Thanks @wzieba! I've tested it and it works as described. Thanks for the fix 🙇.

wzieba committed:
Update test name to match convention
